Src/ILGPU.Algorithms/MatrixOperations/CondensedProductRows.cs

```diff
@@ -0,0 +1,235 @@
+            float[,] a,
+            SparseMatrixShape sparseMatrixShape,
+            T[,] mask,
+            T emptyMaskElement = default)
+            where T : struct, IEquatable<T>
+        {
+            var predicate = new MaskPredicate<T>(mask, emptyMaskElement);
+            return Create(a, sparseMatrixShape, predicate);
+        }
+
+        /// <summary>
+        /// Creates a new condensed representation of a sparse shape and a mask using
+        /// a predicate for each matrix element.
+        /// </summary>
+        /// <param name="a">The dense data input.</param>
+        /// <param name="sparseMatrixShape">The sparse matrix shape.</param>
+        /// <param name="predicate">The masking predicate.</param>
+        /// <returns>The created condensed instance.</returns>
+        public static CondensedProductRows Create<TPredicate>(
+            float[,] a,
+            SparseMatrixShape sparseMatrixShape,
+            TPredicate predicate)
+            where TPredicate : InlineList.IPredicate<Index2D>
+        {
+            int maxNumColumns = sparseMatrixShape.MaxNonZeroEntries;
+            int numSparseRows = sparseMatrixShape.GetLength(0);
+
+            // Count number of non-empty product rows based on Mask && Sparse
+            int numRows = 0;
+            for (int col = 0; col < numSparseRows; ++col)
+            {
+                // We are working with Sparse', so each entry in Sparse.numNeighbors is
+                // a column for A*[column1, column2, ...]
+                int indexLength = sparseMatrixShape.GetNumNeighbors(col);
+                for (int row = 0; row < a.GetLength(0); ++row)
+                {
+                    if (predicate.Apply(new Index2D(row, col)) && indexLength > 0)
+                        ++numRows;
+                }
+            }
+
+            // Allocate storage and copy relevant data
+            var data = new float[numRows, maxNumColumns];
+            var rowIndices = new int[numRows];
+            var columnIndices = new int[numRows];
+            int dataRow = 0;
+
+            // Create the internal data layout
+            Parallel.For(0, numSparseRows, sparseRow =>
+            {
+                int indexLength = sparseMatrixShape.GetNumNeighbors(sparseRow);
+                for (int row = 0; row < sparseMatrixShape.GetLength(0); ++row)
+                {
+                    if (!predicate.Apply(new Index2D(row, sparseRow)))
+                        continue;
+
+                    for (int j = 0; j < indexLength; ++j)
+                    {
+                        int sourceColumn = sparseMatrixShape.GetNeighbor(sparseRow, j);
+                        data[dataRow, j] = a[row, sourceColumn];
+                    }
+
+                    if (indexLength > 0)
+                    {
+                        rowIndices[dataRow] = row;
+                        columnIndices[dataRow] = sparseRow;
+                        ++dataRow;
```

Comment:
Under a parallel for loop the dataRow increment may have problems. I think that maybe we need to acquire a lock, capture the current value, then increment at line 93?

Reply:
We should not use a parallel for in this case anyway...